Implement beforeunload event and infrastructure

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->

Implementation of:
1. https://html.spec.whatwg.org/multipage/#prompt-to-unload-a-document, and
2. https://html.spec.whatwg.org/multipage/#unload-a-document
---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #10787 and fix #20485 and fix #20588 and fix #20496 (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/20329)
<!-- Reviewable:end -->

gterzian committed:
beforeunload and unload infrastructure

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/document.rs, components/script/dom/window.rs, components/script/dom/eventtarget.rs
 * @fitzgen: components/script/dom/document.rs, components/script/dom/window.rs, components/script/dom/eventtarget.rs
 * @kichjang: components/script/dom/document.rs, components/script/dom/window.rs, components/script/dom/eventtarget.rs

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/jdm/web-platform-tests/pull/36.

servo-wpt-sync commented:
Error syncing changes upstream. Logs saved in error-snapshot-1521386860024.

servo-wpt-sync commented:
Error syncing changes upstream. Logs saved in error-snapshot-1521387016450.

servo-wpt-sync commented:
No upstreamable changes; closed existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#36.

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/jdm/web-platform-tests/pull/46.

servo-wpt-sync commented:
No upstreamable changes; closed existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#46.

gterzian committed:
PutForward=href on window and document location

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/jdm/web-platform-tests/pull/47.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#47.

gterzian commented:
@jdm sorry, I will try not to push changes to the tests, unless really meant as as such...

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#47.

gterzian committed:
fire pageshow for new entry after history traversal

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#47.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#47.

servo-wpt-sync commented:
Error syncing changes upstream. Logs saved in error-snapshot-1522682431565.

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/jdm/web-platform-tests/pull/48.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#48.

gterzian commented:
@cbrewster I have one more question, with the current state of this PR, running this test: "tests/wpt/web-platform-tests/html/browsers/browsing-the-web/unloading-documents/beforeunload-on-navigation-of-parent.html", when I change the test to doing a call to `history.go(-2)`, instead of `history.go(-1)`, it passes(it seems that doing only -1 takes you to 'about:blank'). Any ideas?

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#48.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#48.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#48.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20507) made this pull request unmergeable. Please resolve the merge conflicts.

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/jdm/web-platform-tests/pull/49.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#49.

servo-wpt-sync commented:
No upstreamable changes; closed existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#49.

gterzian commented:
@cbrewster Ok this one is good for a first review. 

The tests that are not passing currently, do so because of #20588, #14720, #20496, and #13241, as  well as whatwg/html#3615(I have removed the beforeunload/unload code from History for now).

cbrewster commented on components/script/dom/document.rs:
```diff
@@ -1575,6 +1612,97 @@ impl Document {
         ScriptThread::mark_document_with_no_blocked_loads(self);
     }
 
+    // https://html.spec.whatwg.org/multipage/#unloading-documents
```
nit: https://html.spec.whatwg.org/multipage/#prompt-to-unload-a-document
Label sub-steps

cbrewster commented on components/script/dom/document.rs:
```diff
@@ -1575,6 +1612,97 @@ impl Document {
+            self.salvageable.set(false);
+        }
+        let mut can_unload = true;
+        if event_status == EventStatus::Canceled
```
For now let's remove this if and leave a `TODO` comment.

cbrewster commented on components/script/dom/document.rs:
```diff
@@ -1575,6 +1612,97 @@ impl Document {
+            // TODO: ask the user to confirm that they wish to unload the document.
+        }
+        self.decr_ignore_opens_during_unload_counter();
+        // Step 9
```
We should add `recursive_flag` as a parameter for this method to follow the spec. In most cases it seems the flag is unset. (In fact I can't find a place where this flag is ever set)

cbrewster commented on components/script/dom/document.rs:
```diff
@@ -1575,6 +1612,97 @@ impl Document {
+        self.decr_ignore_opens_during_unload_counter();
+        // Step 9
+        for iframe in self.iter_iframes() {
+            if let Some(document) = iframe.GetContentDocument() {
```
This doesn't handle the case of cross origin iframes, we will probably need to go via the constellation here. This could end up being pretty tricky, I would be okay for this to be a followup.

cbrewster commented on components/script/dom/document.rs:
```diff
@@ -1575,6 +1612,97 @@ impl Document {
+        can_unload
+    }
+
+    // https://html.spec.whatwg.org/multipage/#unloading-documents
```
https://html.spec.whatwg.org/multipage/#unload-a-document
Label steps

cbrewster commented on components/script/dom/document.rs:
```diff
@@ -1575,6 +1612,97 @@ impl Document {
+            let _ = self.window.upcast::<EventTarget>().dispatch_event_with_target(
+                document.root().upcast(),
+                &event,
+            );
```
Add a TODO for step 6: page visibility

gterzian replied:
This one is done as well, right below actually...

cbrewster commented on components/script/dom/document.rs:
```diff
@@ -1575,6 +1612,97 @@ impl Document {
+            self.salvageable.set(!event_handled);
+        }
+        // Step 13
+        for iframe in self.iter_iframes() {
```
Like above, this will need to be done via the constellation in the case of cross origin iframes.

cbrewster commented on components/script/dom/document.rs:
```diff
@@ -1575,6 +1612,97 @@ impl Document {
+        }
+        self.decr_ignore_opens_during_unload_counter();
+        // unloading document cleanup steps.
+        if !self.salvageable.get() {
```
We will need to notify the Constellation to discard the pipeline and set the reloader for the entry to `NeedsToReload::Yes(..)`.

cbrewster commented on components/script/dom/window.rs:
```diff
@@ -1635,8 +1639,12 @@ impl Window {
-        // Suspend timer events.
-        self.upcast::<GlobalScope>().suspend();
+        if let Some(document) = self.document.get() {
+            if !document.salvageable() {
```
I think we will still want to suspend here. AFAICT if the document is not `salvageable` it will be discarded from the session history entirely.

gterzian replied:
For a reason I haven't figured out yet, if a salvageable document is suspended, and resumed, this test still timeouts because the `setTimeout` callback is never called https://github.com/servo/servo/blob/1fe0f32059270e2d3e42b7c7184c2adf56f3f872/tests/wpt/web-platform-tests/html/browsers/browsing-the-web/unloading-documents/unload/006-1.html#L6

cbrewster replied:
Interesting... Something else may be wrong here then. I would still leave this how it was.

gterzian replied:
I've digged a bit deeper, and this seems to be a problem at the level of `timers`, because the document is properly suspended and resumed, but once resumed the timer seems to be ignored because of it's id(see `OneShotTimer.fire_timer`). I'll be debugging it a bit more...

cbrewster replied:
@gterzian interesting, sounds like another followup. Let's open an issue for it and mark the test as failing.

gterzian replied:
Filed https://github.com/servo/servo/issues/20663

cbrewster commented on components/script/dom/window.rs:
```diff
@@ -1651,8 +1659,12 @@ impl Window {
-        // Resume timer events.
-        self.upcast::<GlobalScope>().resume();
+        if let Some(document) = self.document.get() {
+            if document.salvageable() {
```
Ditto

cbrewster commented on components/script/script_thread.rs:
```diff
@@ -2393,6 +2403,19 @@ impl ScriptThread {
             }
         }
 
+        // Step 3 and 4
```
This is only step 4

gterzian replied:
`document.is_unloading` looks at `ignore_opens_during_unload_counter`, which is set by both `unload` and `prompt_to_unload`, so I was thinking this would in fact take care of both steps in one go. Perhaps I should rename the method to `is_unloading_or_prompting`?

cbrewster replied:
Looks like you're right!

cbrewster commented on components/script/timers.rs:
```diff
@@ -235,13 +235,12 @@ impl OneshotTimers {
             return warn!("Suspending an already suspended timer.");
         }
 
-        debug!("Suspending timers.");
```
Why remove this?

cbrewster reviewed (changes requested):
@gterzian Thanks for working on this!
This is a really good start, there are a few concerns I have. Mainly how we handle recursively `prompting to unload` and `unloading`. I think we will need to go via the constellation to do this, it would be nice if we could short circuit on the same origin case.

cc @asajeffrey

gterzian commented on components/constellation/constellation.rs:
```diff
@@ -2415,6 +2410,8 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
             },
             Some(old_pipeline_id) => {
+                // https://html.spec.whatwg.org/multipage/#unload-a-document
+                self.unload_document(old_pipeline_id);
```
@cbrewster Moving the unloading to `change_session_history` seems to work pretty well, with the exception of https://github.com/servo/servo/blob/d8dfa989704591cf6ec2417db6ee3e73530c71de/tests/wpt/web-platform-tests/html/browsers/browsing-the-web/unloading-documents/unload/003.html#L22 which now fails.
It seems that the tests expects the `unload` event to fire while the `iframe.contentDocument` would still be "003-1.html", while the session history is changed when it has already become"002-2.html".

gterzian replied:
I think in this particular case we are unloading as part of [update the session history with the new page](https://html.spec.whatwg.org/multipage/browsing-the-web.html#update-the-session-history-with-the-new-page), and when this should occur is in fact quite vague per the spec, for example "After creating the Document object, but before any script execution, certainly before the parser stops, the user agent must update the session history with the new page." I do think we do it before the parser stops, since `change_session_history` is called in response to receiving a `ScriptMsg::ActivateDocument`, itself sent [here](https://github.com/servo/servo/blob/master/components/script/script_thread.rs#L2175) before the parser is even started. 

Perhaps the test makes too many assumptions about implementation details going beyond the spec?

cbrewster replied:
It is possible that the test is making too many assumptions. I'm okay with looking into this as a followup.

gterzian commented on components/constellation/constellation.rs:
```diff
@@ -2819,7 +2822,19 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
-
+    
+    // Discard the pipeline for a given document, udpdate the joint session history.
+    fn handle_discard_document(&mut self, top_level_browsing_context_id: TopLevelBrowsingContextId, pipeline_id: PipelineId) {
```
@cbrewster I wasn't completely sure what you meant with https://github.com/servo/servo/pull/20329#discussion_r180462129, so I gave it a try here. Please let me know what you think.

gterzian commented:
@cbrewster OK, I think all of your feedback has been addressed, and I have squashed those last commits into one. Perhaps it's ready for a test ride?

bors-servo commented:
:hourglass: Trying commit 34e33072e81664a422a891ddc3eab97e98730b1f with merge 10a34189a700de754cd86e9e575567eedd5e58bd...

gterzian commented:
`command timed out: 1200 seconds without output running ['git', 'fetch', '-t', 'https://github.com/servo/servo', 'try'], attempting to kill
process killed by signal 9
program finished with exit code -1`

cbrewster commented:
Looking at http://build.servo.org/builders/linux-rel-wpt/builds/8525/steps/test__1/logs/stdio
There are a lot of panics:
<details>

```
▶ CRASH [expected OK] /html/browsers/browsing-the-web/navigating-across-documents/javascript-url-abort/javascript-url-abort-return-value-string.tentative.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 17.3.0-devel
  │ No current global object (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at libcore/option.rs:914)
  │ stack backtrace:
  │    0:     0x7fbb339e7fec - backtrace::backtrace::trace::h342f5d76ca5ceb8f
  │    1:     0x7fbb339e73b2 - backtrace::capture::Backtrace::new::hb9e84455e6bd3454
  │    2:     0x7fbb30a2e8ac - servo::main::{{closure}}::h28c771b8a7fc7d4b
  │    3:     0x7fbb33a09039 - std::panicking::rust_panic_with_hook::hfd6818535490a12f
  │                         at libstd/panicking.rs:401
  │    4:     0x7fbb33a08e41 - std::panicking::begin_panic_fmt::hb49e0372029a7978
  │                         at libstd/panicking.rs:347
  │    5:     0x7fbb33a08d92 - rust_begin_unwind
  │                         at libstd/panicking.rs:323
  │    6:     0x7fbb33a2dbc0 - core::panicking::panic_fmt::he07a59762014a101
  │                         at libcore/panicking.rs:71
  │    7:     0x7fbb33a29648 - core::option::expect_failed::h6a663649ad048267
  │                         at libcore/option.rs:914
  │    8:     0x7fbb312efc5b - <script::dom::htmliframeelement::HTMLIFrameElement as script::dom::bindings::codegen::Bindings::HTMLIFrameElementBinding::HTMLIFrameElementBinding::HTMLIFrameElementMethods>::GetContentDocument::hb424f0d55cf57abc
  │    9:     0x7fbb30f011d2 - script::dom::document::Document::unload::h3a38f558f1e39f58
  │   10:     0x7fbb30f32915 - script::script_thread::ScriptThread::handle_msg_from_constellation::h0a23c1b44becdb18
  │   11:     0x7fbb30f2e0d2 - script::script_thread::ScriptThread::handle_msgs::h81bb12819b41d093
  │   12:     0x7fbb30f29ea5 - script::script_thread::ScriptThread::start::h0d6ca36aed57fe5e
  │   13:     0x7fbb314cd592 - profile_traits::mem::ProfilerChan::run_with_memory_reporting::h1ff11c37b9e7e26a
  │   14:     0x7fbb317fb0ee - std::sys_common::backtrace::__rust_begin_short_backtrace::h460025f8ca3ba471
  │   15:     0x7fbb31a817e5 - _ZN3std9panicking3try7do_call17h1a874c3bfbf79f70E.llvm.6011387467664904945
  │   16:     0x7fbb33a1e4fe - __rust_maybe_catch_panic
  │                         at libpanic_unwind/lib.rs:101
  │   17:     0x7fbb319f9950 - <F as alloc::boxed::FnBox<A>>::call_box::h29d76208e37672be
  │   18:     0x7fbb33a12af7 - <alloc::boxed::Box<alloc::boxed::FnBox<A, Output$u3d$R$GT$$u20$$u2b$$u20$$u27$a$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$::call_once::h9987f61f4f93afa5
  │                         at /checkout/src/liballoc/boxed.rs:645
  │                          - std::sys_common::thread::start_thread::h7a671b9cfa13995a
  │                         at libstd/sys_common/thread.rs:24
  │   19:     0x7fbb339fad48 - std::sys::unix::thread::Thread::new::thread_start::had4c521b4cf0110f
  │                         at libstd/sys/unix/thread.rs:90
  │   20:     0x7fbb2f92f183 - start_thread
  │   21:     0x7fbb2e1f703c - clone
  │   22:                0x0 - <unknown>
  │ ERROR 2018-04-18T14:29:27Z: servo: No current global object
  └ Pipeline failed in hard-fail mode.  Crashing!
```
</details>

cbrewster commented on tests/wpt/mozilla/meta/MANIFEST.json:
```diff
@@ -61533,7 +61533,7 @@
   ],
   "css/get-computed-style-for-url.html": [
-   "2e90c0abd6c83bb11113f39a557a4c1c1c24364b",
+   "cb3b467160026d20c4e283c12ded20fe2d025f88",
```
This should be reverted since that test hasn't changed

gterzian commented:
Thanks, I switched to using `document_from_node(&*iframe)` which solved a lot of those panics. I'll have to look into the others. Could you please make it run once more to see if the list was reduced?

cbrewster commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit cc31a8000e200eec4cec8ef27a2d2fc7605ad04a with merge 172d4eb1a84634ccca624f06af83793f900cf53f...

cbrewster commented:
Failures: http://build.servo.org/builders/linux-rel-wpt/builds/8525
<details>

```
{"status": "FAIL", "group": "default", "message": "missing = in const declaration", "stack": "http://web-platform.test:8000/FileAPI/url/resources/fetch-tests.js:64:13", "subtest": "Untitled", "test": "/FileAPI/url/url-with-fetch.any.worker.html", "line": 55000, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: after dispatch expected true got false", "stack": "@http://web-platform.test:8000/dom/events/Event-defaultPrevented-after-dispatch.html:23:5\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1501:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\n@http://web-platform.test:8000/dom/events/Event-defaultPrevented-after-dispatch.html:11:1\n", "subtest": "Event.defaultPrevented is not reset after dipatchEvent()", "test": "/dom/events/Event-defaultPrevented-after-dispatch.html", "line": 62612, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_false: expected false got true", "stack": "@http://web-platform.test:8000/dom/events/EventTarget-dispatchEvent-returnvalue.html:40:5\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1501:20\ntest@http://web-platform.test:8000/resources/testharness.js:511:9\n@http://web-platform.test:8000/dom/events/EventTarget-dispatchEvent-returnvalue.html:23:1\n", "subtest": "Return value of EventTarget.dispatchEvent.", "test": "/dom/events/EventTarget-dispatchEvent-returnvalue.html", "line": 65771, "action": "test_result", "expected": "PASS"}
{"status": "FAIL", "group": "default", "message": "assert_true: expected true got false", "stack": "base_show<@http://web-platform.test:8000/html/browsers/browsing-the-web/unloading-documents/beforeunload-on-navigation-of-parent.html:25:5\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1501:20\nTest.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1525:20\nonpageshow@http://web-platform.test:8000/html/browsers/browsing-the-web/unloading-documents/base.html:11:3\n", "subtest": "beforeunload in iframe on navigation of parent", "test": "/html/browsers/browsing-the-web/unloading-documents/beforeunload-on-navigation-of-parent.html", "line": 66632, "action": "test_result", "expected": "TIMEOUT"}
{"status": "OK", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/browsing-the-web/unloading-documents/beforeunload-on-navigation-of-parent.html", "line": 66635, "action": "test_result", "expected": "TIMEOUT"}
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/browsing-the-web/navigating-across-documents/006.html", "line": 66960, "action": "test_result", "expected": "TIMEOUT"}
{"status": "FAIL", "group": "default", "message": "assert_equals: expected \"href\" but got \"click\"", "stack": "onmessage<@http://web-platform.test:8000/html/browsers/browsing-the-web/navigating-across-documents/007.html:13:5\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1501:20\nTest.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1525:20\n", "subtest": "Link with onclick javascript url and href navigation ", "test": "/html/browsers/browsing-the-web/navigating-across-documents/007.html", "line": 70580, "action": "test_result", "expected": "PASS"}
{"status": "OK", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/browsing-the-web/navigating-across-documents/007.html", "line": 70581, "action": "test_result", "expected": "TIMEOUT"}
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/browsing-the-web/navigating-across-documents/008.html", "line": 70813, "action": "test_result", "expected": "TIMEOUT"}
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_navigate_ancestor-2.html", "line": 90622, "action": "test_result", "expected": "OK"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Document interface: iframe.contentDocument must have own property \"location\"", "test": "/html/dom/interfaces.html", "line": 99984, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Document interface: new Document() must have own property \"location\"", "test": "/html/dom/interfaces.html", "line": 100108, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Document interface: document.implementation.createDocument(null, \"\", null) must have own property \"location\"", "test": "/html/dom/interfaces.html", "line": 100229, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Window interface: window must have own property \"location\"", "test": "/html/dom/interfaces.html", "line": 105574, "action": "test_result", "expected": "FAIL"}
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/browsing-the-web/navigating-across-documents/005.html", "line": 127144, "action": "test_result", "expected": "TIMEOUT"}
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/dom/dynamic-markup-insertion/document-write/write-active-document.html", "line": 129504, "action": "test_result", "expected": "OK"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Returning true with a real iframe unloading; setting returnValue to foo", "test": "/html/browsers/browsing-the-web/unloading-documents/beforeunload-canceling.html", "line": 134550, "action": "test_result", "expected": "TIMEOUT"}
{"status": "OK", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/browsing-the-web/unloading-documents/beforeunload-canceling.html", "line": 134552, "action": "test_result", "expected": "TIMEOUT"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Assignment to location before document is completely loaded", "test": "/html/browsers/history/the-location-interface/assign_before_load.html", "line": 142233, "action": "test_result", "expected": "TIMEOUT"}
{"status": "OK", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/history/the-location-interface/assign_before_load.html", "line": 142234, "action": "test_result", "expected": "TIMEOUT"}
{"status": "FAIL", "group": "default", "message": "assert_equals: expected 3 but got 2", "stack": "do_test<@http://web-platform.test:8000/html/browsers/history/the-location-interface/assign_after_load.html:19:5\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1501:20\nTest.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1525:20\nonload</<@http://web-platform.test:8000/html/browsers/history/the-location-interface/assign_after_load-2.html:5:26\n", "subtest": "Assignment to location after document is completely loaded", "test": "/html/browsers/history/the-location-interface/assign_after_load.html", "line": 146757, "action": "test_result", "expected": "TIMEOUT"}
{"status": "OK", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/history/the-location-interface/assign_after_load.html", "line": 146758, "action": "test_result", "expected": "TIMEOUT"}
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/browsing-the-web/navigating-across-documents/child_navigates_parent_location.html", "line": 147643, "action": "test_result", "expected": "TIMEOUT"}
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/browsing-the-web/navigating-across-documents/javascript-url-abort/javascript-url-abort-return-value-string.tentative.html", "line": 147709, "action": "test_result", "expected": "OK"}
{"status": "FAIL", "group": "default", "message": "missing = in const declaration", "stack": "http://web-platform.test:8000/FileAPI/url/resources/fetch-tests.js:64:13", "subtest": "Untitled", "test": "/FileAPI/url/url-with-xhr.any.worker.html", "line": 164081, "action": "test_result", "expected": "PASS"}
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/browsing-the-web/navigating-across-documents/011.html", "line": 170986, "action": "test_result", "expected": "TIMEOUT"}
{"status": "FAIL", "group": "default", "message": "assert_unreached: Reached unreachable code", "stack": "worker.onerror<@http://web-platform.test:8000/workers/interfaces/WorkerGlobalScope/onerror/handled.html:30:5\nTest.prototype.step@http://web-platform.test:8000/resources/testharness.js:1501:20\nTest.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1525:20\n", "subtest": "onerror, \"handled\"", "test": "/workers/interfaces/WorkerGlobalScope/onerror/handled.html", "line": 189659, "action": "test_result", "expected": "PASS"}
```
</details>

Still getting `No current global object` panics:
<details>

```
  ▶ CRASH [expected TIMEOUT] /html/browsers/browsing-the-web/navigating-across-documents/011.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 17.3.0-devel
  │ No current global object (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at libcore/option.rs:914)
  │ stack backtrace:
  │    0:     0x7fb68488cfec - backtrace::backtrace::trace::h342f5d76ca5ceb8f
  │    1:     0x7fb68488c3b2 - backtrace::capture::Backtrace::new::hb9e84455e6bd3454
  │    2:     0x7fb6818d38ac - servo::main::{{closure}}::h28c771b8a7fc7d4b
  │    3:     0x7fb6848ae039 - std::panicking::rust_panic_with_hook::hfd6818535490a12f
  │                         at libstd/panicking.rs:401
  │    4:     0x7fb6848ade41 - std::panicking::begin_panic_fmt::hb49e0372029a7978
  │                         at libstd/panicking.rs:347
  │    5:     0x7fb6848add92 - rust_begin_unwind
  │                         at libstd/panicking.rs:323
  │    6:     0x7fb6848d2bc0 - core::panicking::panic_fmt::he07a59762014a101
  │                         at libcore/panicking.rs:71
  │    7:     0x7fb6848ce648 - core::option::expect_failed::h6a663649ad048267
  │                         at libcore/option.rs:914
  │    8:     0x7fb682194c5b - <script::dom::htmliframeelement::HTMLIFrameElement as script::dom::bindings::codegen::Bindings::HTMLIFrameElementBinding::HTMLIFrameElementBinding::HTMLIFrameElementMethods>::GetContentDocument::hb424f0d55cf57abc
  │    9:     0x7fb681da61d2 - script::dom::document::Document::unload::h3a38f558f1e39f58
  │   10:     0x7fb681dd7915 - script::script_thread::ScriptThread::handle_msg_from_constellation::h0a23c1b44becdb18
  │   11:     0x7fb681dd30d2 - script::script_thread::ScriptThread::handle_msgs::h81bb12819b41d093
  │   12:     0x7fb681dceea5 - script::script_thread::ScriptThread::start::h0d6ca36aed57fe5e
  │   13:     0x7fb682372592 - profile_traits::mem::ProfilerChan::run_with_memory_reporting::h1ff11c37b9e7e26a
  │   14:     0x7fb6826a00ee - std::sys_common::backtrace::__rust_begin_short_backtrace::h460025f8ca3ba471
  │   15:     0x7fb6829267e5 - _ZN3std9panicking3try7do_call17h1a874c3bfbf79f70E.llvm.6011387467664904945
  │   16:     0x7fb6848c34fe - __rust_maybe_catch_panic
  │                         at libpanic_unwind/lib.rs:101
  │   17:     0x7fb68289e950 - <F as alloc::boxed::FnBox<A>>::call_box::h29d76208e37672be
  │   18:     0x7fb6848b7af7 - <alloc::boxed::Box<alloc::boxed::FnBox<A, Output$u3d$R$GT$$u20$$u2b$$u20$$u27$a$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$::call_once::h9987f61f4f93afa5
  │                         at /checkout/src/liballoc/boxed.rs:645
  │                          - std::sys_common::thread::start_thread::h7a671b9cfa13995a
  │                         at libstd/sys_common/thread.rs:24
  │   19:     0x7fb68489fd48 - std::sys::unix::thread::Thread::new::thread_start::had4c521b4cf0110f
  │                         at libstd/sys/unix/thread.rs:90
  │   20:     0x7fb6807d4183 - start_thread
  │   21:     0x7fb67f09c03c - clone
  │   22:                0x0 - <unknown>
  │ ERROR 2018-04-18T14:30:53Z: servo: No current global object
  └ Pipeline failed in hard-fail mode.  Crashing!
```
</details>

gterzian commented:
@cbrewster Ok, could it be that the bot still used the previous version in running the tests? Since the traceback still refers to `HTMLIFrameElementMethods>::GetContentDocument`, which I replaced with a call to `document_from_node`?
I can imagine the bot grabs the commit via the github API, and that a small lag is possible. 
I tried some of those panicking test 'on my machine', and they also don't panic...

I'll look into some of those PASS/FAIL in the meantime...

gterzian commented:
@cbrewster seems like the "try" didn't come through...

cbrewster commented:
Ugh...
@bors-servo try

cbrewster commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 3fc9b644b6635ac03af0738d264b76c90cd75409 with merge acccc1313c44bf0af1dd67d9756f44a2027e7c28...

gterzian commented:
@cbrewster Could we try again please? Although I think this one might still fail: /html/browsers/browsing-the-web/unloading-documents/beforeunload-on-navigation-of-parent.html

cbrewster commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 2623b326a8ed95f76c5aebfc672a1e1ca63e89e1 with merge 4629070e2b1f6cfb7448c44600bd68e867309590...

gterzian commented:
I'd like you're opinion on this test: https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/html/browsers/browsing-the-web/unloading-documents/beforeunload-on-navigation-of-parent.html

Currently I can't seem to be able to get `beforeunload` to fire in the child iframe of the iframe in the main document, where `history.go` is called. 

I've implemented the `recursiveFlag` as per the spec https://html.spec.whatwg.org/multipage/browsing-the-web.html#prompt-to-unload-a-document Step 9, and it seems to me that it would imply that `beforeunload` would only fire on the first iframe, not the second iframe, since the `recursiveFlag` would seem to prevent such child of an iframe from being prompted.

Is that how you read it too?

cbrewster commented:
@gterzian I wouldn't expect `beforeunload` to be fired there because this PR doesn't handle the history traversal case (notice the `history.go(-1)`), it only handles the navigation case.

gterzian commented:
@cbrewster Ah yes, now I remember we agreed to limit the scope of this PR earlier, which is why I had removed that piece of code. I'll remove it again, and update the test expectation to "FAIL" as opposed to "TIMEOUT".

gterzian commented:
@cbrewster OK I've updated the last test expectation and squashed. 

Regarding that last one, which I've set to FAIL, I'm actually not sure how it was passing previously, it seems that following hyperlinks using "target" hasn't been implemented yet, so I've filed #20673.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20682) made this pull request unmergeable. Please resolve the merge conflicts.

cbrewster commented on components/script/dom/event.rs:
```diff
@@ -482,7 +482,14 @@ fn invoke(window: Option<&Window>,
+        // <https://html.spec.whatwg.org/multipage/#unload-a-document>
+        // Step 9.
+        // Required to establish the 'salvageable' state of document as part of unloading.
+        if event.canceled.get() != EventDefault::Prevented {
```
I don't think this is the right place to do this. See comment below.

cbrewster replied:
Now that I've looked into this more, this doesn't make much sense since we are checking if  `canceled` is `Prevented` and if so, we change it to `Handled`. This shouldn't have an effect since in the `unload` steps you just check if cancelled is not equal to `Allowed`.

gterzian replied:
Yes that makes sense, and the problem is that the default is 'Allowed', to which it is set whether there was a listener handling the event or not(so long as the listener didn't preventDefault), hence the need to specifically mark it as 'Handled' to know the difference between 'no listener triggered', and 'a listener was triggered, but the default was not prevented(and the document is not salvageable)'.

cbrewster commented on components/script/dom/eventtarget.rs:
```diff
@@ -187,22 +187,21 @@ impl CompiledEventListener {
-                                    }
-                                    None => {
-                                        event.upcast::<Event>().PreventDefault();
+                                let rv = event.ReturnValue();
```
https://html.spec.whatwg.org/multipage/webappapis.html#the-event-handler-processing-algorithm Step 5 says we should be setting `E`'s cancelled flag which I believe is done via `mark_as_handled()`.

gterzian replied:
Thanks for looking deeper in that part. It appers there are some things that need clarifying, and after looking more into it as well, here are a few points I think we could consider:

1. Step 5 of "the event handler processing algorithm" writes "Set E's canceled flag", which itself [is defined](https://dom.spec.whatwg.org/#canceled-flag) as a "flag" so either set or unset, and should be set when `preventDefault` is called.
2. In servo, `canceled` [can be either](https://github.com/servo/servo/blob/4f4fb32264ab94406c189471e9ec753a06011b23/components/script/dom/event.rs#L367) `Allowed`, `Prevented`, or `Handled`.
3. calling `PreventDefault` [sets canceled to](https://github.com/servo/servo/blob/4f4fb32264ab94406c189471e9ec753a06011b23/components/script/dom/event.rs#L253) `Prevented`. 
4. `mark_as_handled`  [sets canceled to](https://github.com/servo/servo/blob/4f4fb32264ab94406c189471e9ec753a06011b23/components/script/dom/event.rs#L205) `Handled`. 
5. Step 7 of https://html.spec.whatwg.org/multipage/browsing-the-web.html#prompt-to-unload-a-document and Step 9 of https://html.spec.whatwg.org/multipage/browsing-the-web.html#unload-a-document both write "If any event listeners were triggered", without further defining this concept. 

From the above, I deduct that:

A. Step 5 of "#the-event-handler-processing-algorithm" would in servo translate to "setting `canceled` to `EventDefault::Prevented`(or calling `Event.PreventDefault`).
 I actually missed one part of that step, see the additional commit(to be squashed). 
B. "If any event listeners were triggered" would in servo translate to `canceled` being either `EventDefault::Prevented` or `EventDefault::Handled`, yet it's unclear where that should be done.

I feel like the spec is actually unclear in this regard and could be improved:
1. "If any event listeners were triggered" seems to only relate to the "salvageable" state of the document in the context of (before-)unloading, yet it relates to both "beforeunload" and "unload" events(so it cannot be dealt with only at the level of BeforeUnloadEvent).
2. The spec is missing a general concept of 'the event was handled'.
3. Such a concept could be introduced either as part of [#concept-event-listener-inner-invoke](https://dom.spec.whatwg.org/#concept-event-listener-inner-invoke), which is the implementation I initially went for, or as part of [#the-event-handler-processing-algorithm](https://html.spec.whatwg.org/multipage/webappapis.html#the-event-handler-processing-algorithm)
4. While in servo we could implement this by setting `canceled` to `EventDefault::Handled`, when it wasn't previously set to `EventDefault::Prevented`, in the spec it would probably be clearer to do it as a separate `handled` flag, that could be set at either the end of "#the-event-handler-processing-algorithm" or anytime after Step 1 of "#concept-event-listener-inner-invoke"

What do you think?

cbrewster replied:
I see what you mean with `EventDefault::Handled` vs `EventDefault::Prevented`. It looks like `EventDefault::Handled` is only used when some input events. Not sure why we have `Allowed` `Prevented` and `Handled`, but yes in this case we should just call `PreventDefault` here.

gterzian replied:
Do you mean "mark_as_handled"? 😀 
So we actually need this `Handled`,  to set the document's salvageable state to false when there was any listener triggered for `beforeunload` or `unload`. And I've tried calling `PreventDefault`, and unfortunately that makes a bunch of tests fail that make assertions about wether the default was prevented or not. 

So in another words, we need to mark the event as 'handled', but not 'prevented', hence my somewhat hacky code that you've highligthed. 

Lastly, we need this for the `unload` event as well as `beforeunload`, so it cannot be done only inside `CommonEventHandler::BeforeUnloadEventHandler`(by the way does cover Step 5 "Otherwise", [here](https://github.com/servo/servo/blob/4f4fb32264ab94406c189471e9ec753a06011b23/components/script/dom/eventtarget.rs#L218))

gterzian replied:
asked for input via https://github.com/whatwg/html/issues/3646

gterzian replied:
Ok good news, this has been resolved I think see https://github.com/whatwg/html/issues/3646#issuecomment-384581127
I'll add a commit to deal with this...

cbrewster reviewed (commented):
Almost there! I have one more nit/question.

Thanks again!

gterzian committed:
implement "has event listener", plug into (before)unload

gterzian commented:
Ok this one is ready for another round of review, see the last commit...

gterzian commented:
As a note, I intend to implement "has any listeners" more fully and with additional tests, perhaps together with adding a 'removed' flag to listeners, based on suggestions here https://github.com/whatwg/html/issues/3646#issuecomment-384583278, and I'd rather do it in a follow-up. The current implementation is good enough to make the current tests pass(so it's equivalent to my previous hack).

gterzian commented on components/script/dom/event.rs:
```diff
@@ -102,6 +102,36 @@ impl Event {
+    
+    // Determine if there are any listeners for a given target and type.
+    // See https://github.com/whatwg/dom/issues/453
+    pub fn has_listeners_for(&self, target: &EventTarget, type_: &Atom) -> bool {
```
As a note, I intend to implement "has any listeners" more fully and with additional tests, perhaps together with adding a 'removed' flag to listeners, based on suggestions here whatwg/html#3646 (comment), and I'd rather do it in a follow-up. The current implementation is good enough to make the current tests pass(so it's equivalent to my previous hack).

cbrewster commented on components/script/dom/event.rs:
```diff
@@ -103,6 +103,36 @@ impl Event {
+        // https://dom.spec.whatwg.org/#event-listener-removed
+        let mut event_path = self.construct_event_path(&target);
+        event_path.push(DomRoot::from_ref(target));
+        event_path.iter().filter(|target| target.has_listeners_for(type_)).last().is_some()
```
nit: `event_path.iter().any(|target| target.has_listeners_for(type_))`

cbrewster commented on components/script/dom/event.rs:
```diff
@@ -103,6 +103,36 @@ impl Event {
+
+    // https://dom.spec.whatwg.org/#event-path
+    fn construct_event_path(&self, target: &EventTarget) -> Vec<DomRoot<EventTarget>> {
+        let mut event_path = vec![];
```
In the old code we were doing a `rooted_vec!` to store the elements in the event path, I'm not sure if we can use a `Vec` here.

cc @jdm

jdm replied:
I'm not really clear on why we were using a rooted vector previously. From a safety perspective this change is fine because we're now storing DomRoot values in the vector instead of unrooted Dom values.

gterzian replied:
Ok. By the way we are still using a `rooted_vec!` there: 

https://github.com/servo/servo/pull/20329/files#diff-b9c83174d4ef992d8bc1914c8826e220R165

While I had this new `construct_event_path` return a `Vec<DomRoot<EventTarget>>` so that the caller would still have the option to 'transform' it into a rooted vev if necessary, while this seemed unncessary inside `has_listeners_for`, see https://github.com/servo/servo/pull/20329/files#diff-b9c83174d4ef992d8bc1914c8826e220R111.

Perhaps we can also remove [that call to `rooted_vec!`](https://github.com/servo/servo/pull/20329/files#diff-b9c83174d4ef992d8bc1914c8826e220R165) and simply pass the `Vec<DomRoot<EventTarget>>` to [`dispatch_to_listeners`](https://github.com/servo/servo/pull/20329/files#diff-b9c83174d4ef992d8bc1914c8826e220R167)

cbrewster reviewed (commented):
Almost there!

gterzian commented:
@cbrewster Ok, I've addressed the nit and left the rest as it was... What type of squashing strategy did you have in mind? I was thinking the four commits were somewhat unrelated and perhaps best kept separate, and I'm fine with squashing them if you feel otherwise...

jdm commented:
```
error[E0599]: no method named `replace` found for type `&mut session_history::JointSessionHistory` in the current scope
    --> components/constellation/constellation.rs:3015:14
     |
3015 |             .replace(NeedsToReload::No(pipeline_id), NeedsToReload::Yes(pipeline_id, load_data));
     |              ^^^^^^^
error: aborting due to previous error
```

gterzian commented:
Thanks, rebased off master, and fixed this...

bors-servo commented:
:hourglass: Testing commit 90353ec8676c0902542ceaabf2fdfcef6c4d81f5 with merge 884851b8d85369058abcc1bf45a99d5ff89c3605...

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/jdm/web-platform-tests/pull/57.

gterzian commented on tests/wpt/web-platform-tests/html/browsers/history/joint-session-history/joint-session-history-only-fully-active.html:
```diff
@@ -20,7 +20,7 @@
         // If all documents count, then the grandchild is reachable via the session history,
         // and it has session length 1, so the joint session length change will be 2.
-        assert_equals(2, window.history.length - old_history_len);
+        assert_equals(1, window.history.length - old_history_len);
```
@cbrewster The comment seems to imply that`1` is a valid assertion, actually the one corresponding to only having active documents in the session history, what do you think?

gterzian commented on tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/window-name-after-cross-origin-main-frame-navigation.sub.html.ini:
```diff
@@ -1,3 +1,3 @@
 [window-name-after-cross-origin-main-frame-navigation.sub.html]
   type: testharness
-  expected: TIMEOUT
+  expected: ERROR
```
The error here is actually a servers-side assertion, not a panic, so we went from timeout to fail one could say.

gterzian commented on tests/wpt/metadata/html/browsers/browsing-the-web/history-traversal/window-name-after-same-origin-main-frame-navigation.sub.html.ini:
```diff
@@ -1,3 +1,3 @@
 [window-name-after-same-origin-main-frame-navigation.sub.html]
   type: testharness
-  expected: TIMEOUT
+  expected: ERROR
```
same

gterzian commented on tests/wpt/metadata/html/browsers/windows/auxiliary-browsing-contexts/opener-noreferrer.html.ini:
```diff
@@ -1,5 +1,6 @@
 [opener-noreferrer.html]
   type: testharness
+  expected: TIMEOUT
```
TIMEOUT because we're navigating the main browsing context(while the test calls for navigating '_blank'),   as anchor elements currently ignore their 'target'(it's actually addressed in https://github.com/servo/servo/pull/20678). 

Since this PR properly adds unloading/discarding the document, the test now timeouts because the event handler on the main page is discarded along with the whole page(we should be navigating a new auxiliary browsing context instead).

cbrewster replied:
It looks like this PR is causing the first document to be discarded, which shouldn't happen since, like you said, the first document should be salvageable.

gterzian replied:
Yep, I'm investigating...

gterzian commented:
amended on test, and updated expectations, see my comments for case by case explanation.

servo-wpt-sync commented:
No upstreamable changes; closed existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#57.

gterzian commented on components/script/dom/document.rs:
```diff
@@ -1654,7 +1654,7 @@ impl Document {
         // TODO: Step 6, decrease the event loop's termination nesting level by 1.
         // Step 7
-        self.salvageable.set(has_listeners);
+        self.salvageable.set(!has_listeners);
```
@cbrewster This was the problem, LOL

gterzian commented:
Ok, should be good to go...

cbrewster reviewed (approved):
LGTM! Thanks

cbrewster commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 99eaab4 has been approved by `cbrewster`

<!-- @bors-servo r=cbrewster 99eaab4e559c2ed5dd74db877e82edc61b9d6c26 -->

bors-servo commented:
:hourglass: Testing commit 99eaab4e559c2ed5dd74db877e82edc61b9d6c26 with merge ae80a01af21bc5308c9b300de2b8845e5bd6c079...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11832)

cbrewster commented:
More things passing! :tada: but there is a crash
Failures:

<details>

```
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Document interface: iframe.contentDocument must have own property \"location\"", "test": "/html/dom/interfaces.https.html", "line": 37757, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Document interface: new Document() must have own property \"location\"", "test": "/html/dom/interfaces.https.html", "line": 37878, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Document interface: document.implementation.createDocument(null, \"\", null) must have own property \"location\"", "test": "/html/dom/interfaces.https.html", "line": 37999, "action": "test_result", "expected": "FAIL"}
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "Window interface: window must have own property \"location\"", "test": "/html/dom/interfaces.https.html", "line": 43329, "action": "test_result", "expected": "FAIL"}
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/html/browsers/browsing-the-web/unloading-documents/navigation-within-beforeunload.html", "line": 63084, "action": "test_result", "expected": "OK"}
```
</details>

jdm commented:
```
  ▶ CRASH [expected OK] /html/browsers/browsing-the-web/unloading-documents/navigation-within-beforeunload.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 17.3.0-devel
  │ ERROR 2018-05-04T16:55:30Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(2, PipelineId(0, 1)), pipeline_id: PipelineId(1, 2) }
  │ ERROR 2018-05-04T16:55:30Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(2, PipelineId(0, 1)), pipeline_id: PipelineId(1, 2) }
  │ ERROR 2018-05-04T16:55:30Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(2, PipelineId(0, 1)), pipeline_id: PipelineId(1, 2) }
  │ ERROR 2018-05-04T16:55:30Z: webrender::display_list_flattener: Unknown pipeline used for iframe IframeDisplayItem { clip_id: Clip(2, PipelineId(0, 1)), pipeline_id: PipelineId(1, 4) }
  │ ERROR 2018-05-04T16:55:30Z: script::dom::bindings::error: Error at http://web-platform.test:8000/html/browsers/browsing-the-web/unloading-documents/navigation-within-beforeunload-1.html?0.4542089685414288:7:3 too much recursion
  │ failed to open /dev/urandom: Os { code: 24, kind: Other, message: "Too many open files" } (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(66) }, at libcore/result.rs:945)
  │ 2018-05-04 12:55:31.746 servo[66142:3566091681] Metadata.framework [Error]: couldn't get the client port
  │ stack backtrace:
  │    0:        0x1033c370e - backtrace::backtrace::trace::h7964d766053cc9d7
  │    1:        0x1033c2efc - backtrace::capture::Backtrace::new::h4b3e83f668a53406
  │    2:        0x10085c25e - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::h8065021b052ffffd
  │    3:        0x1033ed564 - std::panicking::rust_panic_with_hook::hcebe05e8466ff134
  │    4:        0x1033ed050 - std::panicking::begin_panic_fmt::hf6867449b867ca07
  │    5:        0x1033ed012 - rust_begin_unwind
  │    6:        0x1033ff315 - core::panicking::panic_fmt::h794f4a50949acedc
  │    7:        0x1033e075d - core::result::unwrap_failed::h3e61bd84938f7a47
  │    8:        0x1033e174a - std::sys::unix::rand::hashmap_random_keys::h4ce375ed6694a11c
  │    9:        0x100c77182 - _$LT$std..collections..hash..map..HashMap$LT$K$C$$u20$V$C$$u20$S$GT$$u20$as$u20$core..default..Default$GT$::default::h9e0ca1059606c377
  │   10:        0x100c3749b - layout_thread::LayoutThread::new::hf1e7a8271e7dbb08
  │   11:        0x100b57d50 - std::sys_common::backtrace::__rust_begin_short_backtrace::h5089b87c9e789e86 (.llvm.4107249418266811869)
  │   12:        0x100bffc6d - std::panicking::try::do_call::ha0c91113a5e277a6 (.llvm.14984580951739245240)
  │   13:        0x1033f8d4e - __rust_maybe_catch_panic
  │   14:        0x100b690a6 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::h70501c75e780521c
  │   15:        0x1033e8ac7 - std::sys_common::thread::start_thread::h3a2a1586806f3926
  │   16:        0x1033d19f8 - std::sys::unix::thread::Thread::new::thread_start::h657bd8405a5d884c
  │   17:     0x7fff95aed059 - _pthread_body
  │   18:     0x7fff95aecfd6 - _pthread_start
  │ ERROR 2018-05-04T16:55:32Z: servo: failed to open /dev/urandom: Os { code: 24, kind: Other, message: "Too many open files" }
  │ called `Result::unwrap()` on an `Err` value: "SendError(..)" (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at libcore/result.rs:945)
  │ stack backtrace:
  │    0:        0x1033c370e - backtrace::backtrace::trace::h7964d766053cc9d7
  │    1:        0x1033c2efc - backtrace::capture::Backtrace::new::h4b3e83f668a53406
  │    2:        0x10085c25e - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::h8065021b052ffffd
  │    3:        0x1033ed564 - std::panicking::rust_panic_with_hook::hcebe05e8466ff134
  │    4:        0x1033ed050 - std::panicking::begin_panic_fmt::hf6867449b867ca07
  │    5:        0x1033ed012 - rust_begin_unwind
  │    6:        0x1033ff315 - core::panicking::panic_fmt::h794f4a50949acedc
  │    7:        0x10175888d - core::result::unwrap_failed::hcf28e664a3a67931
  │    8:        0x101103eba - script::script_thread::ScriptThread::handle_page_headers_available::h45e85421198ed37c
  │    9:        0x100e731b4 - _$LT$std..thread..local..LocalKey$LT$T$GT$$GT$::with::h816675e3a66237c6
  │   10:        0x10108512b - _$LT$script..dom..servoparser..ParserContext$u20$as$u20$net_traits..FetchResponseListener$GT$::process_response::h2e7dd82ccedca151
  │   11:        0x1010fb426 - script::script_thread::ScriptThread::handle_msg_from_constellation::hbf6290441b08a412
  │   12:        0x1010f4432 - script::script_thread::ScriptThread::handle_msgs::h4e232ddff248137d (.llvm.13981216747995023946)
  │   13:        0x101592ad7 - profile_traits::mem::ProfilerChan::run_with_memory_reporting::hdf3bfa49cdb1b55b
  │   14:        0x101833d8a - std::sys_common::backtrace::__rust_begin_short_backtrace::h20e9a09db82a95ed
  │   15:        0x100cfd5bd - std::panicking::try::do_call::h5fbfde9e1429b1d5 (.llvm.2048999851630907103)
  │   16:        0x1033f8d4e - __rust_maybe_catch_panic
  │   17:        0x1019068a6 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::hd582b91943df8e3c
  │   18:        0x1033e8ac7 - std::sys_common::thread::start_thread::h3a2a1586806f3926
  │   19:        0x1033d19f8 - std::sys::unix::thread::Thread::new::thread_start::h657bd8405a5d884c
  │   20:     0x7fff95aed059 - _pthread_body
  │   21:     0x7fff95aecfd6 - _pthread_start
  │ ERROR 2018-05-04T16:55:32Z: servo: called `Result::unwrap()` on an `Err` value: "SendError(..)"
  └ Pipeline failed in hard-fail mode.  Crashing!
```

jdm commented:
The crash from mac-dev-unit is not caused by these changes; it's clearly something wrong lower down in the Rust standard library:
```
  * frame #0: 0x000000010063a8f2 text_util-9a954535af80505d`std::panicking::panicking::hbf60d20c3ea01bc0 [inlined] core::ptr::swap_nonoverlapping_bytes::h5ceec518bedbc8b5 + 28 at ptr.rs:237 [opt]
    frame #1: 0x000000010063a8d6 text_util-9a954535af80505d`std::panicking::panicking::hbf60d20c3ea01bc0 [inlined] core::ptr::swap_nonoverlapping::h135f714f781a3d4a at ptr.rs:187 [opt]
    frame #2: 0x000000010063a8d6 text_util-9a954535af80505d`std::panicking::panicking::hbf60d20c3ea01bc0 [inlined] core::mem::swap::ha8a5fb0906ce1619 at mem.rs:634 [opt]
    frame #3: 0x000000010063a8d6 text_util-9a954535af80505d`std::panicking::panicking::hbf60d20c3ea01bc0 [inlined] core::mem::replace::hc5999657bde25f00 at mem.rs:691 [opt]
    frame #4: 0x000000010063a8d6 text_util-9a954535af80505d`std::panicking::panicking::hbf60d20c3ea01bc0 [inlined] _$LT$std..thread..local..LocalKey$LT$T$GT$$GT$::init::h03c530c70809d764 at local.rs:270 [opt]
    frame #5: 0x000000010063a8d6 text_util-9a954535af80505d`std::panicking::panicking::hbf60d20c3ea01bc0 [inlined] _$LT$std..thread..local..LocalKey$LT$T$GT$$GT$::try_with::h1ddab9e8e4e02c71 + 34 at local.rs:296 [opt]
    frame #6: 0x000000010063a8b4 text_util-9a954535af80505d`std::panicking::panicking::hbf60d20c3ea01bc0 [inlined] _$LT$std..thread..local..LocalKey$LT$T$GT$$GT$::with::h7ad9abd4ee126f73 at local.rs:248 [opt]
    frame #7: 0x000000010063a8b4 text_util-9a954535af80505d`std::panicking::panicking::hbf60d20c3ea01bc0 [inlined] std::panicking::update_panic_count::h8ff9c302601a529d at panicking.rs:240 [opt]
    frame #8: 0x000000010063a8b4 text_util-9a954535af80505d`std::panicking::panicking::hbf60d20c3ea01bc0 + 4 at panicking.rs:317 [opt]
    frame #9: 0x0000000100019363 text_util-9a954535af80505d`std::sys_common::backtrace::__rust_begin_short_backtrace::h5ce7af51dff78437 [inlined] std::thread::panicking::ha8ec11437d5dda69 + 563 at mod.rs:650 [opt]
    frame #10: 0x000000010001935e text_util-9a954535af80505d`std::sys_common::backtrace::__rust_begin_short_backtrace::h5ce7af51dff78437 [inlined] std::sys_common::poison::Flag::borrow::h4e2e930fbcd29fe9 at poison.rs:36 [opt]
    frame #11: 0x000000010001935e text_util-9a954535af80505d`std::sys_common::backtrace::__rust_begin_short_backtrace::h5ce7af51dff78437 [inlined] _$LT$std..sync..mutex..MutexGuard$LT$$u27$mutex$C$$u20$T$GT$$GT$::new::hf958532833c3eeee at mutex.rs:426 [opt]
    frame #12: 0x000000010001935e text_util-9a954535af80505d`std::sys_common::backtrace::__rust_begin_short_backtrace::h5ce7af51dff78437 [inlined] _$LT$std..sync..mutex..Mutex$LT$T$GT$$GT$::lock::h08aa85ea884e0a79 + 9 at mutex.rs:231 [opt]
    frame #13: 0x0000000100019355 text_util-9a954535af80505d`std::sys_common::backtrace::__rust_begin_short_backtrace::h5ce7af51dff78437 [inlined] test::run_test::run_test_inner::_$u7b$$u7b$closure$u7d$$u7d$::h293891f148241b3f + 508 at lib.rs:1414 [opt]
    frame #14: 0x0000000100019159 text_util-9a954535af80505d`std::sys_common::backtrace::__rust_begin_short_backtrace::h5ce7af51dff78437 + 41 at backtrace.rs:136 [opt]
    frame #15: 0x0000000100030388 text_util-9a954535af80505d`std::panicking::try::do_call::h317793a359ef8eae [inlined] std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::hce2a129863ba037b + 40 at mod.rs:409 [opt]
    frame #16: 0x0000000100030372 text_util-9a954535af80505d`std::panicking::try::do_call::h317793a359ef8eae [inlined] _$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$::call_once::ha38aec20aaf8023b at panic.rs:305 [opt]
    frame #17: 0x0000000100030372 text_util-9a954535af80505d`std::panicking::try::do_call::h317793a359ef8eae + 18 at panicking.rs:310 [opt]
    frame #18: 0x000000010064657f text_util-9a954535af80505d`__rust_maybe_catch_panic + 31 at lib.rs:105 [opt]
    frame #19: 0x000000010001ac55 text_util-9a954535af80505d`_$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::hbda32c8bc14400fa [inlined] std::panicking::try::hd50d66bdaa9b80f6 + 57 at panicking.rs:289 [opt]
    frame #20: 0x000000010001ac1c text_util-9a954535af80505d`_$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::hbda32c8bc14400fa [inlined] std::panic::catch_unwind::h23f85cf47b3beb69 + 7 at panic.rs:374 [opt]
    frame #21: 0x000000010001ac15 text_util-9a954535af80505d`_$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::hbda32c8bc14400fa [inlined] std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::hb9992794009509d0 + 55 at mod.rs:408 [opt]
    frame #22: 0x000000010001abde text_util-9a954535af80505d`_$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::hbda32c8bc14400fa + 46 at boxed.rs:638 [opt]
    frame #23: 0x0000000100637758 text_util-9a954535af80505d`std::sys_common::thread::start_thread::h3a2a1586806f3926 [inlined] _$LT$alloc..boxed..Box$LT$alloc..boxed..FnBox$LT$A$C$$u20$Output$u3d$R$GT$$u20$$u2b$$u20$$u27$a$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$::call_once::he272aed47e96b55c + 136 at boxed.rs:648 [opt]
    frame #24: 0x0000000100637755 text_util-9a954535af80505d`std::sys_common::thread::start_thread::h3a2a1586806f3926 + 133 at thread.rs:24 [opt]
    frame #25: 0x0000000100629689 text_util-9a954535af80505d`std::sys::unix::thread::Thread::new::thread_start::h657bd8405a5d884c + 9 at thread.rs:90 [opt]
    frame #26: 0x00007fff8afbb05a libsystem_pthread.dylib`_pthread_body + 131
    frame #27: 0x00007fff8afbafd7 libsystem_pthread.dylib`_pthread_start + 176
    frame #28: 0x00007fff8afb83ed libsystem_pthread.dylib`thread_start + 13
```

gterzian commented on components/script/dom/window.rs:
```diff
@@ -1550,18 +1550,33 @@ impl Window {
-            MainThreadScriptMsg::Navigate(pipeline_id,
-                LoadData::new(url, Some(pipeline_id), referrer_policy, Some(doc.url())), replace)).unwrap();
+        
+        // Step 4
```
Moved [Step 4 #navigating-across-documents](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigating-across-documents) from `script_thread`(previously this was done in response to receiving the `MainThreadScriptMsg::Navigate` sent a bit below) to here, which fits the spec steps better actually.

gterzian commented:
I've done https://github.com/servo/servo/pull/20329#discussion_r186254422, which prevents the crash inside `/html/browsers/browsing-the-web/unloading-documents/navigation-within-beforeunload.html`, and still marking the test as expected `FAIL`. 

Having read the spec a bit more closely, and the test, I think this test probably doesn't match the spec anymore(it was added back in https://github.com/w3c/web-platform-tests/pull/304). 

That is because the test seems to rely on an infinite loop where the navigation inside the `beforeunload` callback triggers another `beforeunload` event which triggers another callback and so on. See https://github.com/servo/servo/blob/8e978dab17e18ff62a9eda3089ba2b072cd2429b/tests/wpt/web-platform-tests/html/browsers/browsing-the-web/unloading-documents/navigation-within-beforeunload-1.html

It seems that [Step 4 #navigating-across-documents](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigating-across-documents) precludes such an loop.

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8690)

KiChjang commented:
Looks like we have a new passing test!

```
{"status": "PASS", "group": "default", "message": null, "stack": null, "subtest": "font-weight transition", "test": "/css/css-fonts/variations/font-weight-interpolation.html", "line": 49827, "action": "test_result", "expected": "TIMEOUT"}
```

KiChjang commented:
Wait sorry, false alarm, that's #20760. I don't think beforeunloadevent affects that test...

cbrewster commented:
@bors-servo retry
 * #20734
 * #20760

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/11480), [arm32](http://build.servo.org/builders/arm32/builds/11711), [arm64](http://build.servo.org/builders/arm64/builds/11757), [linux-dev](http://build.servo.org/builders/linux-dev/builds/11795), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8694), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11842), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/4243), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/4269), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8511), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/4338), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/7188) are reusable. Rebuilding only [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8690), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8572), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/4270)...

gterzian commented:
@cbrewster thanks for your help!
